File: C++/Merge_sort.cpp
// بسم الله الرحمن الرحيم
// "وَأَنْ لَيْسَ لِلْإِنسَانِ إِلَّا مَا سَعَىٰ"
#include <bits/stdc++.h>
using namespace std;
#define sz(x) (int)x.size()
#define nl '\n'

// Time complexity : O( n log(n) )
void Merge(vector<int> &left, vector<int> &right, vector<int> &nums) // Merge section
{
    int ls = sz(nums) / 2, rs = sz(nums) - ls;
    int i = 0, l = 0, r = 0;
    while (l < ls && r < rs)
    {
        if (left[l] < right[r])
            nums[i] = left[l], i++, l++;
        else
            nums[i] = right[r], i++, r++;
    }
    while (l < ls)
        nums[i] = left[l], i++, l++;

    while (r < rs)
        nums[i] = right[r], i++, r++;
}

void Merge_sort(vector<int> &nums) // divide section
{
    int n = sz(nums);
    if (n <= 1)
        return;

    int mid = n / 2;
    vector<int> left(mid), right(n - mid);
    for (int i = 0, j = 0; i < n; i++)
    {
        if (i < mid)
            left[i] = nums[i];
        else
            right[j] = nums[i], j++;
    }

    Merge_sort(left);
    Merge_sort(right);
    Merge(left, right, nums);
}

int main()
{
    int n;
    cin >> n;
    vector<int> nums(n);
    for (int i = 0; i < n; i++)
        cin >> nums[i];

    Merge_sort(nums);

    for (int i = 0; i < n; i++)
        cout << nums[i] << " ";

    cout << nl;

    return 0;
}

File: Python/Merge_sort.py
# بسم الله الرحمن الرحيم
# "وَأَنْ لَيْسَ لِلْإِنسَانِ إِلَّا مَا سَعَىٰ"


# Time complexity: O(n long(n) )
def Merge(left, right, nums):
    ls = len(left)
    rs = len(right)
    i = l = r = 0
    while l < ls and r < rs:
        if left[l] < right[r]:
            nums[i] = left[l]
            l += 1
        else:
            nums[i] = right[r]
            r += 1
        i += 1

    while l < ls:
        nums[i] = left[l]
        i += 1
        l += 1

    while r < rs:
        nums[i] = right[r]
        i += 1
        r += 1


def Merge_sort(nums):
    n = len(nums)
    if n <= 1:
        return
    mid = n // 2
    left = [0] * mid
    right = [0] * (n - mid)

    j = 0
    for i in range(n):
        if i < mid:
            left[i] = nums[i]
        else:
            right[j] = nums[i]
            j += 1

    Merge_sort(left)
    Merge_sort(right)
    Merge(left, right, nums)


def main():
    n = int(input())
    nums = list(map(int, input().split()))
    Merge_sort(nums)
    print(*nums)


if __name__ == "__main__":
    main()
